sycl/plugins/esimd_cpu/pi_esimd_cpu.cpp

```diff
@@ -115,121 +177,208 @@ LAMBDA_WRAPPER_TMPL(sycl::nd_item<3>, NDITEM_3DIM, 3)
-  // info retrieved from CM with GlobalOffset/GlobalSize/LocalSize
-  // info and call Lambda function
-  // LambdaWrapper->Func(nd_item_1dim);
+  const sycl::group<1> Group = IDBuilder::createGroup<1>(
+      LambdaWrapper->GlobalSize, LambdaWrapper->LocalSize, GroupSize, GroupID);
+
+  const sycl::id<1> GlobalID = GroupID * LambdaWrapper->LocalSize + LocalID +
+                               LambdaWrapper->GlobalOffset;
+  const sycl::item<1, /*Offset=*/true> GlobalItem =
+      IDBuilder::createItem<1, true>(LambdaWrapper->GlobalSize, GlobalID,
+                                     LambdaWrapper->GlobalOffset);
+  const sycl::item<1, /*Offset=*/false> LocalItem =
+      IDBuilder::createItem<1, false>(LambdaWrapper->LocalSize, LocalID);
+
+  const sycl::nd_item<1> NDItem1Dim =
+      IDBuilder::createNDItem<1>(GlobalItem, LocalItem, Group);
+
+  LambdaWrapper->Func(NDItem1Dim);
 }
 
-extern "C" inline void invokeLambda_NDITEM_2DIM(void *Wrapper) {
+inline void InvokeLambda_NDITEM_2DIM(void *Wrapper) {
   auto *LambdaWrapper = reinterpret_cast<LambdaWrapper_NDITEM_2DIM *>(Wrapper);
+  sycl::range<2> GroupSize(
+      sycl::detail::InitializedVal<2, sycl::range>::template get<0>());
+
+  for (int I = 0; I < 2 /*Dims*/; ++I) {
+    if (LambdaWrapper->LocalSize[I] == 0 ||
+        LambdaWrapper->GlobalSize[I] % LambdaWrapper->LocalSize[I] != 0) {
+      throw sycl::nd_range_error("Invalid local size for global size - 2DIM",
+                                 PI_INVALID_WORK_GROUP_SIZE);
+    }
+    GroupSize[I] = LambdaWrapper->GlobalSize[I] / LambdaWrapper->LocalSize[I];
+  }
+
+  const sycl::id<2> LocalID = {cm_support::get_thread_idx(0),
+                               cm_support::get_thread_idx(1)};
+
+  const sycl::id<2> GroupID = {cm_support::get_group_idx(0),
+                               cm_support::get_group_idx(1)};
+
+  const sycl::group<2> Group = IDBuilder::createGroup<2>(
+      LambdaWrapper->GlobalSize, LambdaWrapper->LocalSize, GroupSize, GroupID);
 
-  PLACEHOLDER_UNUSED(LambdaWrapper);
-  // TODO : construct cl::sycl::nd_item<2> instance using thread ID
-  // info retrieved from CM with GlobalOffset/GlobalSize/LocalSize
-  // info and call Lambda function
-  // LambdaWrapper->Func(nd_item_2dim);
+  const sycl::id<2> GlobalID = GroupID * LambdaWrapper->LocalSize + LocalID +
+                               LambdaWrapper->GlobalOffset;
+  const sycl::item<2, /*Offset=*/true> GlobalItem =
+      IDBuilder::createItem<2, true>(LambdaWrapper->GlobalSize, GlobalID,
+                                     LambdaWrapper->GlobalOffset);
+  const sycl::item<2, /*Offset=*/false> LocalItem =
+      IDBuilder::createItem<2, false>(LambdaWrapper->LocalSize, LocalID);
+
+  const sycl::nd_item<2> NDItem2Dim =
+      IDBuilder::createNDItem<2>(GlobalItem, LocalItem, Group);
+
+  LambdaWrapper->Func(NDItem2Dim);
 }
 
-extern "C" inline void invokeLambda_NDITEM_3DIM(void *Wrapper) {
+inline void InvokeLambda_NDITEM_3DIM(void *Wrapper) {
```

Comment:
suggest to template all `InvokeLambda*` and `LambdaWrapper*` by the number of dimensions and reduce the code by 2/3.

```cpp
template <int NDims> struct RangeBuilder;

template <> struct RangeBuilder<1> {
  template <typename Gen> static range<1> create(Gen G) { return range<1>{ G(0)}; }
};
template <> struct RangeBuilder<2> {
  template <typename Gen> static range<2> create(Gen G) { return range<2>{ G(0), G(1) }; }
};
template <> struct RangeBuilder<3> {
  template <typename Gen> static range<3> create(Gen G) { return range<3>{ G(0), G(1), G(3) }; }
};

template <int NDims>
void InvokeLambda_NDITEM(void *Wrapper) {
  auto *LambdaWrapper = reinterpret_cast<LambdaWrapper_NDITEM<NDims>*>(Wrapper);
  sycl::range<NDims> GroupSize(
    sycl::detail::InitializedVal<NDims, sycl::range>::template get<0>());

  for (int I = 0; I < NDims /*Dims*/; ++I) {
    if (LambdaWrapper->LocalSize[I] == 0 ||
      LambdaWrapper->GlobalSize[I] % LambdaWrapper->LocalSize[I] != 0) {
      throw sycl::nd_range_error("Invalid local size for global size - 3DIM",
        PI_INVALID_WORK_GROUP_SIZE);
    }
    GroupSize[I] = LambdaWrapper->GlobalSize[I] / LambdaWrapper->LocalSize[I];
  }

  const sycl::id<NDims> LocalID = RangeBuilder<NDims>::create(
    [](int i) { return cm_support::get_thread_idx(i); }
  );

  const sycl::id<NDims> GroupID = RangeBuilder<NDims>::create(
    [](int i) { return cm_support::get_group_idx(i); }
  );

  const sycl::group<NDims> Group = IDBuilder::createGroup<NDims>(
    LambdaWrapper->GlobalSize, LambdaWrapper->LocalSize, GroupSize, GroupID);

  const sycl::id<NDims> GlobalID = GroupID * LambdaWrapper->LocalSize + LocalID +
    LambdaWrapper->GlobalOffset;
  const sycl::item<NDims, /*Offset=*/true> GlobalItem =
    IDBuilder::createItem<NDims, true>(LambdaWrapper->GlobalSize, GlobalID,
      LambdaWrapper->GlobalOffset);
  const sycl::item<NDims, /*Offset=*/false> LocalItem =
    IDBuilder::createItem<NDims, false>(LambdaWrapper->LocalSize, LocalID);

  const sycl::nd_item<NDims> NDItem3Dim =
    IDBuilder::createNDItem<NDims>(GlobalItem, LocalItem, Group);

  LambdaWrapper->Func(NDItem3Dim);
}
```

Reply:
> `LambdaWrapper*` functions are prefixed with `extern "C"`. Can I apply template programming for such functions?

Disregard this comment. I'm working on templated functions as suggested.

Reply:
I'll apply suggested refactoring after the ESIMD_CPU feature becomes fully working in sycl branch.

Reply:
Suggested template is applied.